pkg/controllers/management/multiclusterapp/handler_multiclusterapp.go

```diff
@@ -144,47 +145,30 @@ func (mc *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.
 			if rt.Context == "project" {
 				prtbName = mcapp.Name + "-" + r
 				// check if these prtbs already exist
-				_, err := mc.prtbLister.Get(projectName, prtbName)
-				if err != nil {
-					if apierrors.IsNotFound(err) {
-						_, err := mc.prtbs.Create(&v3.ProjectRoleTemplateBinding{
-							ObjectMeta: v1.ObjectMeta{
-								Name:      prtbName,
-								Namespace: projectName,
-							},
-							UserName:         systemUser.Name,
-							RoleTemplateName: r,
-							ProjectName:      p.ProjectName,
-						})
-						if err != nil && !apierrors.IsAlreadyExists(err) {
-							return nil, err
-						}
-					} else {
-						return nil, err
-					}
+				if err = mc.createPrtb(projectName, p.ProjectName, prtbName, r, systemUser.Name, systemUserPrincipalID); err != nil {
+					return nil, err
 				}
 			} else if rt.Context == "cluster" {
 				crtbName = mcapp.Name + "-" + r
-				// check if these crtbs already exist
-				_, err := mc.crtbLister.Get(clusterName, crtbName)
-				if err != nil {
-					if apierrors.IsNotFound(err) {
-						_, err := mc.crtbs.Create(&v3.ClusterRoleTemplateBinding{
-							ObjectMeta: v1.ObjectMeta{
-								Name:      crtbName,
-								Namespace: clusterName,
-							},
-							UserName:         systemUser.Name,
-							RoleTemplateName: r,
-							ClusterName:      clusterName,
-						})
-						if err != nil && !apierrors.IsAlreadyExists(err) {
-							return nil, err
-						}
-					} else {
-						return nil, err
-					}
-				}
+				mc.createCrtb(clusterName, crtbName, r, systemUser.Name, systemUserPrincipalID)
+			}
+		}
+	}
+
+	for target, roles := range mcapp.Spec.TargetToRole {
+		split := strings.SplitN(target, ":", 2)
+		if len(split) == 2 {
+			// create prtbs with these roles
+			projectName := split[1]
+			for _, r := range roles {
+				prtbName = mcapp.Name + "-" + r
+				mc.createPrtb(projectName, target, prtbName, r, systemUser.Name, systemUserPrincipalID)
+			}
+		} else {
+			// create crtbs with the roles
+			for _, r := range roles {
+				crtbName = mcapp.Name + "-" + r
+				mc.createCrtb(target, crtbName, r, systemUser.Name, systemUserPrincipalID)
```

Comment:
@mrajashree Can you please explain this change?

Reply:
@prachidamle in the Store Create method, when no Roles are passed, I'm saving the creator's derived roles in TargetToRoles map, where key could be the projectId or the clusterId.
ProjectID will be of the form "c1:p1", and cluster id will only be "c1". Hence this if-else. And based on it I'm either creating prtb or crtb. Does this sound good?